power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/model.hpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_MODEL_HPP
+#define POWER_GRID_MODEL_CPP_MODEL_HPP
+
+#include "power_grid_model_c/model.h"
+
+#include "basics.hpp"
+#include "dataset.hpp"
+#include "handle.hpp"
+#include "options.hpp"
+
+namespace power_grid_model_cpp {
+class Model {
+  public:
+    // constructor
+    Model(double system_frequency, DatasetConst const& input_dataset)
+        : model_{PGM_create_model(handle_.get(), system_frequency, input_dataset.get())} {}
+    // copy constructor
+    Model(Model const& other) : model_{PGM_copy_model(handle_.get(), other.model_.get())} {
+        other.handle_.check_error();
+    }
+    // copy assignment operator
+    Model& operator=(Model const& other) {
+        if (this != &other) {
+            model_.reset(PGM_copy_model(handle_.get(), other.model_.get()));
+            other.handle_.check_error();
+        }
+        return *this;
+    }
+    // destructor
+    ~Model() = default;
```

Comment:
rule of 3/5/0: if you specify one of them, you should specify all of them
```suggestion
    Model(Model const& other) : model_{PGM_copy_model(handle_.get(), other.model_.get())} {
        other.handle_.check_error();
    }
    Model& operator=(Model const& other) {
        if (this != &other) {
            model_.reset(PGM_copy_model(handle_.get(), other.model_.get()));
            other.handle_.check_error();
        }
        return *this;
    }
    Model(Model&& other) = default;
    Model& operator=(Model&& other) = default;
    ~Model() = default;
```